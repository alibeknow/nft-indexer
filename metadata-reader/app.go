package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
