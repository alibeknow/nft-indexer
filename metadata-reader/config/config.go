package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
