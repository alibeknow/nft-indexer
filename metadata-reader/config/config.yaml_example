server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
