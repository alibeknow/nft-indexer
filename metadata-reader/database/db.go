package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
