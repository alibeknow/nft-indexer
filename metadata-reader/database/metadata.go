package database

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"metadata-reader/metadata"
)

type MetadataDB struct {
	collection *mongo.Collection
}

func (mDB MetadataDB) InsertOrUpdate(ctx context.Context, m metadata.Metadata) error {
	currentTime := time.Now()

	filter := bson.M{"_id": m.ID}
	update := bson.M{
		"$set": bson.M{
			"metadata":   m.Metadata,
			"type":       m.Type,
			"updated_at": currentTime,
		},
		"$setOnInsert": bson.M{
			"created_at": currentTime,
		},
	}

	opts := options.Update().SetUpsert(true)
	_, err := mDB.collection.UpdateOne(ctx, filter, update, opts)
	return err
}
