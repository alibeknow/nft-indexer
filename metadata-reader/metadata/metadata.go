package metadata

import (
	"context"
	"time"
)

type Metadata struct {
	ID        string     `bson:"_id"`
	Metadata  string     `bson:"metadata"`
	Type      string     `bson:"type"`
	CreatedAt *time.Time `bson:"created_at,omitempty"`
	UpdatedAt *time.Time `bson:"updated_at,omitempty"`
}

type DB interface {
	InsertOrUpdate(ctx context.Context, m Metadata) error
}
