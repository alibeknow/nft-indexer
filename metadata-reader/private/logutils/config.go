package logutils

type Config struct {
	Level string
}
