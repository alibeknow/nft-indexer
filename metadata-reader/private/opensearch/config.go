package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
