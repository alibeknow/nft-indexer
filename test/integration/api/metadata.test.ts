process.env.METADATA_REPO_PROVIDER = 'mongo';

import { getApiConfig } from '@api/app.config';
import { MetadataModule } from '@api/metadata';
import { HttpServer, INestApplication } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { Test } from '@nestjs/testing';
import { Blockchain } from '@shared/blockchain';
import { ContractsRepo } from '@shared/contracts';
import { DB, DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { Metadata, MetadataRepoDB } from '@shared/metadata';
import { TokenStandard, TokensRepo } from '@shared/tokens';
import { LoggerModule } from 'nestjs-pino';
import nock from 'nock';
import request from 'supertest';
import { getNestApp } from '../../helpers/app';
import { fixtureMetdata } from '../fixtures/fixtures.metadata';
import { fixtureApiTokens } from '../fixtures/token.api.fixture';
import { staxResponseFixture } from './fixtures';

async function seedMetadataRepo(db: DBClass): Promise<(Metadata | null)[]> {
  const metadataRepo = new MetadataRepoDB(db);
  const contractRepo = new ContractsRepo(db);
  const tokenRepo = new TokensRepo(db);
  await contractRepo.insertIfNone({
    blockchain: Blockchain.ETH,
    address: '0xE1E484251ffFee048A839bc2d68C160BD8C82aBB',
    block: 15177280,
    name: 'Lost Illusion',
    type: TokenStandard.ERC721,
    createdAt: new Date(),
  });
  await contractRepo.insertIfNone({
    blockchain: Blockchain.ETH,
    address: '0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f',
    block: 15177280,
    name: 'Lost Illusion',
    type: TokenStandard.ERC721,
    createdAt: new Date(),
  });

  await Promise.all(
    fixtureApiTokens.map((token) => {
      const [ , contractAddress, tokenId ] = token._id.split(':');
      tokenRepo.save({
        block: token.block,
        blockchain: Blockchain.ETH,
        from: '0x95d35C8a511F5877Af28B515ab4f0A03B730ae9C',
        contractAddress,
        tokenId,
        count: token.count,
        createdAt: token.createdAt,
        tokenUri: token.tokenUri,
      });
    }),
  );

  return Promise.all(
    fixtureMetdata.map(({ _id, metadata, type }) =>
      metadataRepo.save(_id, metadata, type),
    ),
  );
}

jest.setTimeout(100000);

describe('[Metadata endpoints]', () => {
  let app: INestApplication;
  let httpServer: HttpServer;
  let db: DBClass;

  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `api-metadata-test-${dbConfig.db}` });
    await db.open();

    await db.metadata().deleteMany({});
    await db.tokens().deleteMany({});
    await db.contracts().deleteMany({});
    await seedMetadataRepo(db);

    const moduleRef = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          load: [ getApiConfig ],
          isGlobal: true,
        }),
        MetadataModule,
        LoggerModule.forRoot(),
      ],
    })
      .overrideProvider(DB)
      .useValue(db)
      .compile();

    app = await getNestApp(moduleRef);
    httpServer = app.getHttpServer();

    nock('https://test.stax')
      .persist()
      .get(/.*/)
      .reply(200, { staxResponseFixture });
  });

  describe('[GET /api/v1/:chainName/metadata/:contractAddress/:tokenId]', () => {
    it('should return 404 error if requested item doesn`t exist in our DB', async () => {
      const actualResult = await request(httpServer)
        .get(
          '/api/v1/eth/metadata/0x33EeCbf908478C10614626A9D304bfe18B78DD73/0x74727e18',
        )
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(404);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return 400 error if validation doesn`t pass', async () => {
      const actualResult = await request(httpServer)
        .get(
          '/api/v0/solana/metadata/0x33EeCbf908478C10614626A9D304bfe18B78DD73/0x74727e18',
        )
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(400);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct result based on request params', async () => {
      const actualResult1 = await request(httpServer)
        .get(
          '/api/v1/eth/metadata/0xE1E484251ffFee048A839bc2d68C160BD8C82aBB/0x0441',
        )
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
    });

    it('should return correct result based on request params even if a contract address is not checksum formatted', async () => {
      const actualResult1 = await request(httpServer)
        .get(
          `/api/v1/eth/metadata/${'0xE1E484251ffFee048A839bc2d68C160BD8C82aBB'.toLowerCase()}/0x0441`,
        )
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
    });
  });

  describe('[GET /api/v0/:chainName/metadata/:contractAddress]', () => {
    it('should return 400 error if requests doesn`t pass validation', async () => {
      const actualResult1 = await request(httpServer)
        .get('/api/v0/eth/metadata/0x33EeCbf908478C10614626A9D304bfe18B78DD73')
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get(
          '/api/v0/solana/metadata/0x33EeCbf908478C10614626A9D304bfe18B78DD73?tokenIds=0x5b',
        )
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(400);
      expect(actualResult1.body).toMatchSnapshot();
      expect(actualResult2.statusCode).toBe(400);
      expect(actualResult2.body).toMatchSnapshot();
    });

    it('should return correct results based on params and query params', async () => {
      const actualResult1 = await request(httpServer)
        .get(
          '/api/v0/eth/metadata/0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f?tokenIds=0x0340,0x0341',
        )
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get(
          '/api/v0/eth/metadata/0x0e28B28681B1C6b58561A97720F50176ECA387E5?tokenIds=0x2612',
        )
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body).toMatchSnapshot();
    });

    it('should return correct results based on params and query params even if contract address is not checksum formatted', async () => {
      const actualResult1 = await request(httpServer)
        .get(
          `/api/v0/eth/metadata/${'0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f'.toLowerCase()}?tokenIds=0x0340,0x0341`,
        )
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get(
          '/api/v0/eth/metadata/0X0e28B28681B1C6b58561A97720f50176eca387E5?tokenIds=0x2612',
        )
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);

      expect(actualResult1.body).toMatchSnapshot();
      expect(actualResult2.statusCode).toBe(200);

      expect(actualResult2.body).toMatchSnapshot();
    });
  });

  afterAll(async () => {
    nock.cleanAll();
    await db.close(true);
    await app.close();
  });
});
