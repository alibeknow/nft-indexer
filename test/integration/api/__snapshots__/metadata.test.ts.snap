// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Metadata endpoints] [GET /api/v0/:chainName/metadata/:contractAddress] should return 400 error if requests doesn\`t pass validation 1`] = `
{
  "error": "Bad Request",
  "message": [
    "tokenIds must be an array",
    "tokenIds should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`[Metadata endpoints] [GET /api/v0/:chainName/metadata/:contractAddress] should return 400 error if requests doesn\`t pass validation 2`] = `
{
  "error": "Bad Request",
  "message": [
    "chainName must be a valid enum value",
  ],
  "statusCode": 400,
}
`;

exports[`[Metadata endpoints] [GET /api/v0/:chainName/metadata/:contractAddress] should return correct results based on params and query params 1`] = `
{
  "result": [],
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [GET /api/v0/:chainName/metadata/:contractAddress] should return correct results based on params and query params 2`] = `
{
  "result": [
    {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "Orange",
        },
        {
          "trait_type": "Body",
          "value": "Body",
        },
        {
          "trait_type": "Beak",
          "value": "Upset",
        },
        {
          "trait_type": "Eyes",
          "value": "Scared",
        },
        {
          "trait_type": "Eyebrows",
          "value": "Default",
        },
        {
          "trait_type": "Trousers",
          "value": "Pink",
        },
        {
          "trait_type": "Shirt",
          "value": "None",
        },
        {
          "trait_type": "Feet",
          "value": "Smart Shoes",
        },
        {
          "trait_type": "Hat",
          "value": "Green Bucket Hat",
        },
        {
          "trait_type": "Item",
          "value": "Pancakes",
        },
      ],
      "compiler": "Engine",
      "content": [
        {
          "type": "image",
          "url": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
        },
      ],
      "date": 1643342687198,
      "description": "The star from AlphaBetty Books first publication.",
      "dna": "8763393910575a1668ec6696617204632dd80db9",
      "edition": 9745,
      "image": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
      "ledger_metadata": {
        "ledger_stax_image": "{"staxResponseFixture":{"name":"NFT Token - Stax Metadata","image":"https://some.image.com/ipfs/123","describe":"my fascinating image"}}",
      },
      "name": "The Seagull finds his talent #9745",
    },
  ],
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [GET /api/v0/:chainName/metadata/:contractAddress] should return correct results based on params and query params even if contract address is not checksum formatted 1`] = `
{
  "result": [],
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [GET /api/v0/:chainName/metadata/:contractAddress] should return correct results based on params and query params even if contract address is not checksum formatted 2`] = `
{
  "result": [
    {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "Orange",
        },
        {
          "trait_type": "Body",
          "value": "Body",
        },
        {
          "trait_type": "Beak",
          "value": "Upset",
        },
        {
          "trait_type": "Eyes",
          "value": "Scared",
        },
        {
          "trait_type": "Eyebrows",
          "value": "Default",
        },
        {
          "trait_type": "Trousers",
          "value": "Pink",
        },
        {
          "trait_type": "Shirt",
          "value": "None",
        },
        {
          "trait_type": "Feet",
          "value": "Smart Shoes",
        },
        {
          "trait_type": "Hat",
          "value": "Green Bucket Hat",
        },
        {
          "trait_type": "Item",
          "value": "Pancakes",
        },
      ],
      "compiler": "Engine",
      "content": [
        {
          "type": "image",
          "url": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
        },
      ],
      "date": 1643342687198,
      "description": "The star from AlphaBetty Books first publication.",
      "dna": "8763393910575a1668ec6696617204632dd80db9",
      "edition": 9745,
      "image": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
      "ledger_metadata": {
        "ledger_stax_image": "{"staxResponseFixture":{"name":"NFT Token - Stax Metadata","image":"https://some.image.com/ipfs/123","describe":"my fascinating image"}}",
      },
      "name": "The Seagull finds his talent #9745",
    },
  ],
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [GET /api/v1/:chainName/metadata/:contractAddress/:tokenId] should return 400 error if validation doesn\`t pass 1`] = `
{
  "error": "Bad Request",
  "message": [
    "chainName must be a valid enum value",
  ],
  "statusCode": 400,
}
`;

exports[`[Metadata endpoints] [GET /api/v1/:chainName/metadata/:contractAddress/:tokenId] should return 404 error if requested item doesn\`t exist in our DB 1`] = `
{
  "error": "Not Found",
  "message": "Contract address '0x33EeCbf908478C10614626A9D304bfe18B78DD73' not found",
  "statusCode": 404,
}
`;

exports[`[Metadata endpoints] [GET /api/v1/:chainName/metadata/:contractAddress/:tokenId] should return correct result based on request params 1`] = `
{
  "result": {
    "chain": "eth",
    "contractMetadata": {
      "address": "0xE1E484251ffFee048A839bc2d68C160BD8C82aBB",
      "name": "Lost Illusion",
      "tokenType": "ERC-721",
    },
    "metadata": {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "background-96",
        },
        {
          "trait_type": "Flare",
          "value": "Light red",
        },
        {
          "trait_type": "Clothes",
          "value": "Striped vest",
        },
        {
          "trait_type": "Ears",
          "value": "Normal",
        },
        {
          "trait_type": "Mouth",
          "value": "Tongue out",
        },
        {
          "trait_type": "Eyes",
          "value": "Laser eye",
        },
        {
          "trait_type": "Top",
          "value": "Party",
        },
      ],
      "description": "Rich ape #1088 from unique collection, find out more at https://www.richapes.club",
      "image": "ipfs://QmRYe4DGMqBUquW6yzHFRFBNMU794ygRrxvFBwLJ6xfArG/1088.png",
      "ledger_metadata": {
        "ledger_stax_image": "{"staxResponseFixture":{"name":"NFT Token - Stax Metadata","image":"https://some.image.com/ipfs/123","describe":"my fascinating image"}}",
      },
    },
    "tokenId": "0x0441",
    "tokenType": "ERC-721",
    "tokenURI": "ipfs://QmdqggiNGU1oxhjpn6t8GUrtCXSSuVmdbTz1FhtRCcnFyy/382.json",
  },
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [GET /api/v1/:chainName/metadata/:contractAddress/:tokenId] should return correct result based on request params even if a contract address is not checksum formatted 1`] = `
{
  "result": {
    "chain": "eth",
    "contractMetadata": {
      "address": "0xE1E484251ffFee048A839bc2d68C160BD8C82aBB",
      "name": "Lost Illusion",
      "tokenType": "ERC-721",
    },
    "metadata": {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "background-96",
        },
        {
          "trait_type": "Flare",
          "value": "Light red",
        },
        {
          "trait_type": "Clothes",
          "value": "Striped vest",
        },
        {
          "trait_type": "Ears",
          "value": "Normal",
        },
        {
          "trait_type": "Mouth",
          "value": "Tongue out",
        },
        {
          "trait_type": "Eyes",
          "value": "Laser eye",
        },
        {
          "trait_type": "Top",
          "value": "Party",
        },
      ],
      "description": "Rich ape #1088 from unique collection, find out more at https://www.richapes.club",
      "image": "ipfs://QmRYe4DGMqBUquW6yzHFRFBNMU794ygRrxvFBwLJ6xfArG/1088.png",
      "ledger_metadata": {
        "ledger_stax_image": "{"staxResponseFixture":{"name":"NFT Token - Stax Metadata","image":"https://some.image.com/ipfs/123","describe":"my fascinating image"}}",
      },
    },
    "tokenId": "0x0441",
    "tokenType": "ERC-721",
    "tokenURI": "ipfs://QmdqggiNGU1oxhjpn6t8GUrtCXSSuVmdbTz1FhtRCcnFyy/382.json",
  },
  "statusCode": 200,
}
`;
